src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    internal sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
+    {
+        public override string ConvertName(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+                return name;
+
+            // Allocates a string builder with the guessed result length,
+            // where 5 is the average word length in English, and
+            // max(2, length / 5) is the number of underscores.
+            StringBuilder builder = new StringBuilder(name.Length + Math.Max(2, name.Length / 5));
+            UnicodeCategory? previousCategory = null;
+
+            for (int currentIndex = 0; currentIndex < name.Length; currentIndex++)
+            {
+                char currentChar = name[currentIndex];
+                if (currentChar == '_')
+                {
+                    builder.Append('_');
+                    previousCategory = null;
+                    continue;
+                }
+
+                UnicodeCategory currentCategory = char.GetUnicodeCategory(currentChar);
+
+                switch (currentCategory)
+                {
+                    case UnicodeCategory.UppercaseLetter:
+                    case UnicodeCategory.TitlecaseLetter:
+                        if (previousCategory == UnicodeCategory.SpaceSeparator ||
+                            previousCategory == UnicodeCategory.LowercaseLetter ||
+                            previousCategory != UnicodeCategory.DecimalDigitNumber &&
+                            currentIndex > 0 &&
+                            currentIndex + 1 < name.Length &&
+                            char.IsLower(name[currentIndex + 1]))
+                        {
+                            builder.Append('_');
+                        }
+
+                        currentChar = char.ToLower(currentChar);
+                        break;
+
+                    case UnicodeCategory.LowercaseLetter:
+                    case UnicodeCategory.DecimalDigitNumber:
```

Comment:
What's the rational for special casing `DecimalDigitNumber`? Can you share/add a test case?

Reply:
This category contains a bunch of characters. Is 0-9 sufficient?
https://www.fileformat.info/info/unicode/category/Nd/list.htm

Reply:
Tests already exists for numbers in names. One of them you have used in benchmarks (:

Since the policy should put underscores and not remove any characters and numbers including non-ASCII, this category should be used.

Reply:
Ah yes, I see the test case in the bug https://github.com/npgsql/npgsql/issues/2152.

`ABC123 -> abc123` clearly indicates the need for using 0-9 (`ab_c123` is clearly wrong). I see the numbers tests now :)

To be explicit, having non-ASCII digit tests would be good too (i.e. pick others from the Nd list).

Reply:
What's about non-decimal numbers, math symbols and other characters? Maybe invert the logic and write everything except punctuation and spaces?

/cc @GrabYourPitchforks

Reply:
> everything except punctuation and spaces

That makes sense to me.

Not sure how accurate this is, but it is a data point (from https://capitalizemytitle.com/camel-case/):
> Basic Snake Case Capitalization Rules
All letters are lowercase.
All spaces between words are filled with underscores.
Remove all punctuation.